zanata-war/etc/scripts/Mysql2Postgresql.groovy

```diff
@@ -0,0 +1,479 @@
+    def mysqlschema = options.mysqlschema
+    def mysqluser = options.mysqluser ?: ""
+    def mysqlpassword = options.mysqlpassword ?: ""
+
+    def psqlhost = options.psqlhost ?: "localhost"
+    def psqlport = options.psqlport ?: "5432"
+    def psqldb = options.psqldb
+    def psqluser = options.psqluser ?: ""
+    def psqlpassword = options.psqlpassword ?: ""
+
+    if(!mysqlschema) {
+        println "You must provide a mysql schema to migrate from"
+        cli.usage()
+        return false
+    }
+    if(!psqldb && !generateDDL) {
+        println "You must provide a PostgreSql schema to migrate into"
+        cli.usage()
+        return false
+    }
+
+    mysqldb = Sql.newInstance("jdbc:mysql://${mysqlhost}:${mysqlport}/${mysqlschema}",
+        "${mysqluser}", "${mysqlpassword}", "com.mysql.jdbc.Driver")
+    mysqlMetadata = mysqldb.connection.metaData
+    // tell mysql driver to conserve memory (streaming resultset)
+    mysqldb.resultSetConcurrency = ResultSet.CONCUR_READ_ONLY
+    mysqldb.withStatement{stmt -> stmt.fetchSize = Integer.MIN_VALUE }
+
+    def count = mysqldb.firstRow('''
+        select count(*) as count from DATABASECHANGELOG
+        where filename='db/changelogs/db.changelog-3.7.xml\'
+        and author = 'aeng@redhat.com\'
+        and id = '6\'''').count
+    if (count == 0) {
+        println "ERROR: database is too old for migration to postgresql"
+        println "database must be updated to Zanata 3.7.1+ first"
+        return false
+    }
+
+    if(generateDDL) {
+        // Nothing to do
+    }
+    else {
+        postgresdb = Sql.newInstance(
+            "jdbc:postgresql://${psqlhost}:${psqlport}/${psqldb}",
+            "${psqluser}", "${psqlpassword}", "org.postgresql.Driver")
+    }
+    return true
+}
+
+// Writes the statement on a destination, which could be a DDL file or a
+// PotgreSQL database.
+def executeOnDestination(String statement) {
+    // Write to filepsq /res
+    if(generateDDL) {
+        // Append a delimiter, if not already present
+        if(!statement.trim().endsWith(";")) {
+            statement = statement + ";"
+        }
+        ddlFile << statement + "\n"
+        ddlFile.flush()
+    }
+    // Write to postgresql
+    else {
+        postgresdb.execute(statement)
+    }
+}
+
+def createTable(String tableName, ResultSet columns) {
+    println "Creating table $tableName"
+    StringBuilder createTableSql = new StringBuilder("CREATE TABLE $tableName ( ")
+    boolean firstCol = true
+    while(columns.next()) {
+        def colName = columns.getString(4)
+        def dataType = columns.getInt(5)
+        def dataTypeName = columns.getString(6)
+        def size = columns.getInt(7)
+        def isNullable = "YES".equalsIgnoreCase( columns.getString(18) ) ? true : false
+        def isAutoIncrement = "YES".equalsIgnoreCase( columns.getString(23) ) ? true : false
+        def isGenerated = "YES".equalsIgnoreCase( columns.getString(24) ) ? true : false
```

Comment:
Can we refer to them by name instead of position? Failing that, we should
at least define some constants with meaningful names.

On 25 September 2015 at 16:28, Carlos Munoz <notifications@github.com>
wrote:

> In zanata-war/etc/scripts/Mysql2Postgresql.groovy
> <https://github.com/zanata/zanata-server/pull/938#discussion_r40402686>:
>
> > +        postgresdb.execute(statement)
> > +    }
> > +}
> > +
> > +def createTable(String tableName, ResultSet columns) {
> > +    println "Creating table $tableName"
> > +    StringBuilder createTableSql = new StringBuilder("CREATE TABLE $tableName ( ")
> > +    boolean firstCol = true
> > +    while(columns.next()) {
> > +        def colName = columns.getString(4)
> > +        def dataType = columns.getInt(5)
> > +        def dataTypeName = columns.getString(6)
> > +        def size = columns.getInt(7)
> > +        def isNullable = "YES".equalsIgnoreCase( columns.getString(18) ) ? true : false
> > +        def isAutoIncrement = "YES".equalsIgnoreCase( columns.getString(23) ) ? true : false
> > +        def isGenerated = "YES".equalsIgnoreCase( columns.getString(24) ) ? true : false
>
> Those [image: :sparkles:] magic numbers [image: :sparkles:] come from:
>
> http://docs.oracle.com/javase/7/docs/api/java/sql/DatabaseMetaData.html#getColumns(java.lang.String,%20java.lang.String,%20java.lang.String,%20java.lang.String)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/zanata/zanata-server/pull/938/files#r40402686>.
>



-- 
Sean Flanigan

Principal Software Engineer
Globalisation Tools Engineering
Red Hat